internal/k8s/reconciler/listener.go

```diff
@@ -119,12 +125,85 @@ func (l *K8sListener) validateTLS(ctx context.Context) error {
 		}
 		l.status.ResolvedRefs.InvalidCertificateRef = certificateErr
 	} else {
-		l.certificates = []string{resource}
+		l.tlsParams.Certificates = []string{resource}
+	}
+
+	if l.listener.TLS.Options != nil {
+		tlsMinVersion := l.listener.TLS.Options[tlsMinVersionAnnotationKey]
+		tlsMaxVersion := l.listener.TLS.Options[tlsMaxVersionAnnotationKey]
+		tlsCipherSuitesStr := l.listener.TLS.Options[tlsCipherSuitesAnnotationKey]
+
+		if tlsMinVersion != "" {
+			if _, ok := supportedTlsVersions[string(tlsMinVersion)]; !ok {
+				l.status.Ready.Invalid = errors.New("unrecognized TLS min version")
+				return nil
+			}
+
+			if tlsCipherSuitesStr != "" {
+				if _, ok := tlsVersionsWithConfigurableCipherSuites[string(tlsMinVersion)]; !ok {
+					l.status.Ready.Invalid = errors.New("configuring TLS cipher suites is only supported for TLS 1.2 and earlier")
+					return nil
+				}
+			}
+
+			l.tlsParams.MinVersion = string(tlsMinVersion)
+		}
+
+		if tlsMaxVersion != "" {
+			if _, ok := supportedTlsVersions[string(tlsMaxVersion)]; !ok {
+				l.status.Ready.Invalid = errors.New("unrecognized TLS max version")
+				return nil
+			}
+
+			l.tlsParams.MaxVersion = string(tlsMaxVersion)
+		}
+
+		if tlsCipherSuitesStr != "" {
+			// split comma delimited string into string array and trim whitespace
+			tlsCipherSuitesUntrimmed := strings.Split(string(tlsCipherSuitesStr), ",")
+			tlsCipherSuites := tlsCipherSuitesUntrimmed[:0]
+			for _, c := range tlsCipherSuitesUntrimmed {
+				tlsCipherSuites = append(tlsCipherSuites, strings.TrimSpace(c))
+			}
+
+			// validate each cipher suite in array
+			for _, c := range tlsCipherSuites {
+				if _, ok := supportedTlsCipherSuites[c]; !ok {
+					l.status.Ready.Invalid = fmt.Errorf("unrecognized or unsupported TLS cipher suite: %s", c)
+					return nil
+				}
+			}
+
+			// set cipher suites on listener TLSParams
+			l.tlsParams.CipherSuites = tlsCipherSuites
+		}
 	}
 
 	return nil
 }
 
+// FIXME: import this from Consul API
+var supportedTlsVersions = map[string]struct{}{
+	"TLS_AUTO": {},
+	"TLSv1_0":  {},
+	"TLSv1_1":  {},
+	"TLSv1_2":  {},
+	"TLSv1_3":  {},
+}
+
+// FIXME: import this from Consul API
+var tlsVersionsWithConfigurableCipherSuites = map[string]struct{}{
+	"TLS_AUTO": {}, // Remove if Envoy ever sets TLS 1.3 as default minimum
+	"TLSv1_0":  {},
+	"TLSv1_1":  {},
+	"TLSv1_2":  {},
+}
+
+// FIXME: import this from Consul API
+var supportedTlsCipherSuites = map[string]struct{}{
+	"TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256": {},
```

Comment:
If this setting the default cipher suites enabled on the Listener? If so, the listener needs a much larger list than this for the default cipher suites.

Reply:
Yes, this was just a placeholder for testing - I was hoping to have a single source of truth for this list exposed through the Consul API but that doesn't seem like it will be possible. In the future it would be nice to have this list as part of a common xDS library used by both Consul core and Consul API Gateway.